### PR TITLE
Use wasm-dce to tree shake WASM

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
no

**If relevant, link to documentation update:**
n/a

**Summary**
Use wasm-dce

**Does this PR introduce a breaking change?**
no

**Other information**

### CODE REVIEW

**Comment by sokra** [comment]
What happens to Custom Section while rewriting?

**Comment by xtuc** [comment]
They will be removed. Note that this is not wanted, we are working on it.

**Comment by sokra** [comment]
I don't like how this converts the WASM to text format first...

**Comment by xtuc** [comment]
I don't like it either, we are working on emitting wasm directly (custom sections are not even support in WAST yet).

**Comment by xtuc** [comment]
Ok @webpack-bot, i'll have to fix this.

File: lib/WebAssemblyGenerator.js [hunk 1]
```diff
@@ -4,8 +4,16 @@
...
 
+const wasmDce = require("wasm-dce");
+const RawSource = require("webpack-sources").RawSource;
+
 class WebAssemblyGenerator {
 	generate(module) {
+		if(Array.isArray(module.usedExports)) {
+			// rewrite WASM
+			const buf = module.originalSource().source();
+			return new RawSource(wasmDce(buf, module.usedExports));
```
Comment (ooflorent) [inline lib/WebAssemblyGenerator.js]: Maybe this feature should be behind a flag to keep the dev builds fast

**Comment by xtuc** [comment]
These errors are related to the fact that we're printing WAST and then converting to WASM.

It's better to print WASM directly, that will require a bit of work and #6531 is priority I think. We can leave this PR open I can PR against it once we're ready.

**Comment by xtuc** [comment]
We're not ready with `wasm-dce` now, I'll resubmit a PR later.

**Pull request closed without merging**